File: Neetcode-150-js/Stack/valid-parentheses.js

// https://leetcode.com/problems/valid-parentheses/description/

/**
 * @param {string} s
 * @return {boolean}
 */


const OPENING = ['(', '{', '['];


Array.prototype.empty = function(){
    return this.length === 0;
}

Array.prototype.back = function(){
    return this[this.length-1];
}

var isValid = function(s) {
    const stack = [];

    for(let i = 0; i < s.length; i++){
        let curr = s[i];
        
        if(OPENING.includes(curr)) stack.push(curr);
        else{
            if(stack.empty()) return false;
            else if(curr === ')' && stack.back() !== '(') return false;
            else if(curr === ']' && stack.back() !== '[') return false;
            else if(curr === '}' && stack.back() !== '{') return false;
            else stack.pop();
        }
    }

    if(!stack.empty()) return false;
    return true;
};

File: Striver Important Interview Questions/Day 13 - Stack and Queues/valid-parentheses.cpp

// https://leetcode.com/problems/valid-parentheses/description/

class Solution {
public:
    bool isValid(string s) {
        
        stack<char> st;
        for(int i = 0; i < s.size(); i++){
            char ch = s[i];
            if(ch == '(' || ch == '[' || ch == '{') st.push(ch);
            else{
                if(st.empty()) return false;
                else if(ch == ']' && st.top() == '['){
                    st.pop();
                    continue;
                }
                else if(ch == '}' && st.top() == '{'){
                    st.pop();
                    continue;
                }
                else if(ch == ')' && st.top() == '('){
                    st.pop();
                    continue;
                }else return false;
            }
        }

        if(st.empty()) return true;
        return false;
    }
};